src/sentry/utils/sentry_apps/webhooks.py

```diff
@@ -126,69 +127,87 @@ def send_and_save_webhook_request(
+            response_code=response.status_code,
             org_id=org_id,
             event=event,
             url=url,
+            error_id=response.headers.get("Sentry-Hook-Error"),
+            project_id=response.headers.get("Sentry-Hook-Project"),
+            response=response,
             headers=app_platform_event.headers,
         )
-        record_timeout(sentry_app, org_id, e)
-        # Re-raise the exception because some of these tasks might retry on the exception
-        raise
-
-    track_response_code(response.status_code, slug, event)
-    buffer.add_request(
-        response_code=response.status_code,
-        org_id=org_id,
-        event=event,
-        url=url,
-        error_id=response.headers.get("Sentry-Hook-Error"),
-        project_id=response.headers.get("Sentry-Hook-Project"),
-        response=response,
-        headers=app_platform_event.headers,
-    )
-    # we don't disable alert rules for internal integrations
-    # so we don't want to consider responses related to them
-    # for the purpose of disabling integrations
-    if app_platform_event.action != "event.alert":
-        record_response_for_disabling_integration(sentry_app, org_id, response)
-
-    if response.status_code == status.HTTP_503_SERVICE_UNAVAILABLE:
-        raise ApiHostError.from_request(response.request)
-
-    elif response.status_code == status.HTTP_504_GATEWAY_TIMEOUT:
-        raise ApiTimeoutError.from_request(response.request)
-
-    elif 400 <= response.status_code < 500:
-        raise ClientError(response.status_code, url, response=response)
-
-    response.raise_for_status()
-    return response
+        # we don't disable alert rules for internal integrations
+        # so we don't want to consider responses related to them
+        # for the purpose of disabling integrations
+        if app_platform_event.action != "event.alert":
+            record_response_for_disabling_integration(sentry_app, org_id, response)
+
+        if response.status_code == status.HTTP_503_SERVICE_UNAVAILABLE:
+            lifecycle.record_halt(
+                halt_reason=f"send_and_save_webhook_request.{SentryAppWebhookHaltReason.INTEGRATOR_ERROR}"
+            )
+            raise ApiHostError.from_request(response.request)
+
+        elif response.status_code == status.HTTP_504_GATEWAY_TIMEOUT:
+            lifecycle.record_halt(
+                halt_reason=f"send_and_save_webhook_request.{SentryAppWebhookHaltReason.INTEGRATOR_ERROR}"
+            )
+            raise ApiTimeoutError.from_request(response.request)
+
+        elif 400 <= response.status_code < 500:
+            lifecycle.record_halt(
+                halt_reason=f"send_and_save_webhook_request.{SentryAppWebhookHaltReason.GOT_CLIENT_ERROR}_{response.status_code}"
+            )
+            raise ClientError(response.status_code, url, response=response)
+
+        try:
+            response.raise_for_status()
+        except RequestException as e:
```

Comment:
why is `RequestException` a halt?

Reply:
Everything after we've sent a webhook we don't care about. Getting a RequestException here means we got a 500 - 600 response code from our webhook, and we don't consider that a fail for sending webhooks.